dali/kernels/imgproc/convolution/convolution_gpu.h

```diff
@@ -0,0 +1,175 @@
+                    in_shape.size(), " vs ", window_size.size(), "."));
+    int num_samples = in_shape.size();
+    for (int i = 0; i < num_samples; i++) {
+      int num_channels = has_channels ? in_shape[i][ndim - 1] : 1;
+      DALI_ENFORCE(window_size[i][0] % 2 == 1,
+                   make_string("Kernel window should have odd length, got: ", window_size,
+                               " for sample ", i, "."));
+
+      DALI_ENFORCE((window_size[i][0] / 2) * num_channels < kMaxRadiusSpan,
+                   make_string("Kernel window too big for sample ", i, "."));
+    }
+    se.add<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    se.add<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+    req.scratch_sizes = se.sizes;
+    req.output_shapes.push_back(in_shape);
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim> out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const TensorListView<StorageCPU, const W, 1>& window, W scale = 1) {
+    int num_samples = in.size();
+    auto* window_tmp_buffer_host =
+        ctx.scratchpad->Allocate<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    auto* window_tmp_buffer_gpu =
+        ctx.scratchpad->Allocate<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+
+    // Pad and align windows
+    AlignWindows(window_tmp_buffer_host, window, in.shape);
+    cudaMemcpyAsync(window_tmp_buffer_gpu, window_tmp_buffer_host,
+                    sizeof(W) * num_samples * kWindowCopyBufferSize, cudaMemcpyHostToDevice,
+                    ctx.gpu.stream);
+
+    Arguments args;
+    if (kInnerConv) {
+      // Inner (or rather innermost) - repack arguments
+      for (int i = 0; i < num_samples; i++) {
+        cutlass::Array<int, 2> size;
+        auto sample_shape = in.tensor_shape(i);
+        int num_channels = has_channels ? sample_shape[ndim - 1] : 1;
+        // height
+        size[0] = volume(sample_shape.begin(), sample_shape.begin() + axis);
+        // width
+        size[1] = sample_shape[ndim - has_channels - 1];
+        int row_stride = sample_shape[ndim - has_channels - 1] * num_channels;
+        auto* window_gpu = window_tmp_buffer_gpu + i * kWindowCopyBufferSize;
+        args.push_back(SampleArguments{
+            size,                                              // Input matrix dimensions
+            static_cast<int>(window.tensor_shape_span(i)[0]),  // Window sizes
+            num_channels,                                      // channels count (innermost)
+            {in.tensor_data(i), row_stride},                   // Tensor-ref for source matrix A
+            window_gpu,                                        // Pointers to windows
+            {out.tensor_data(i), row_stride},                  // Tensor-ref for source matrix C
+            {out.tensor_data(i), row_stride},  // Tensor-ref for destination matrix D
```

Comment:
You can align comments.
Also, what is the difference between `source matrix C` and ` destination matrix D`?

Reply:
I trimmed that comment that just said it can be the same memory, so not really that helpful :man_shrugging: 

It's the `D = alpha * A*B + beta * C`. The C and D can be the same matrix.

Reply:
Maybe you can write (I think it is not listed anywhere) what is the naming convention: `D = alpha * A*B + beta * C`, and put a remark that this happens in place.

Reply:
I extended a top doxygen about what is happening in this kernel, I can move it if you want. Will post the changes when I fix everything.

Reply:
Doxygen should probably do.